odl/discr/tensor_ops.py

```diff
@@ -0,0 +1,599 @@
+        if self.is_weighted:
+            out *= self.weights[0]
+
+        if self.domain.size == 1:
+            return
+
+        tmp = self.range.element()
+        for fi, w in zip(vf[1:], self.weights[1:]):
+            fi.ufunc.absolute(out=tmp)
+            if self.is_weighted:
+                tmp *= w
+            out.ufunc.maximum(tmp, out=out)
+
+    def _call_vecfield_p(self, vf, out):
+        """Implement ``self(vf, out)`` for exponent 1 < p < ``inf``."""
+        # Optimization for 1 component - just absolute value (maybe weighted)
+        if self.domain.size == 1:
+            vf[0].ufunc.absolute(out=out)
+            if self.is_weighted:
+                out *= self.weights[0] ** (1 / self.exponent)
+            return
+
+        # Initialize out, avoiding one copy
+        self._abs_pow_ufunc(vf[0], out=out)
+        if self.is_weighted:
+            out *= self.weights[0]
+
+        tmp = self.range.element()
+        for fi, w in zip(vf[1:], self.weights[1:]):
+            self._abs_pow_ufunc(fi, out=tmp)
+            if self.is_weighted:
+                tmp *= w
+            out += tmp
+
+        out.ufunc.power(1 / self.exponent, out=out)
+
+    def _abs_pow_ufunc(self, fi, out):
+        """Compute |F_i(x)|^p point-wise and write to ``out``."""
+        # Optimization for a very common case
+        if self.exponent == 2.0 and self.base_space.field == RealNumbers():
+            out.multiply(fi, fi)
+        else:
+            fi.ufunc.absolute(out=out)
+            out.ufunc.power(self.exponent, out=out)
+
+    def derivative(self, vf):
+        """Derivative of the point-wise norm operator at ``vf``.
+
+        The derivative at ``F`` of the point-wise norm operator ``N``
+        with finite exponent ``p`` and weights ``w`` is the pointwise
+        inner product with the vector field
+
+            ``x --> N(F)(x)^(1-p) * [ F_j(x) * |F_j(x)|^(p-2) ]_j``.
+
+        Note that this is not well-defined for ``F = 0``. If ``p < 2``,
+        any zero component will result in a singularity.
+
+        Parameters
+        ----------
+        vf : domain `element-like`
+            Vector field ``F`` at which to evaluate the derivative
+
+        Returns
+        -------
+        deriv : `PointwiseInner`
+            Derivative operator at the given point ``vf``
+
+        Raises
+        ------
+        NotImplementedError
+            * if the vector field space is complex, since the derivative
+              is not linear in that case
+            * if the exponent is ``inf``
+        """
+        if self.domain.field == ComplexNumbers():
+            raise NotImplementedError('operator not Frechet-differentiable '
+                                      'on a complex space.')
+
+        if self.exponent == float('inf'):
+            raise NotImplementedError('operator not Frechet-differentiable '
+                                      'for exponent = inf.')
+
+        vf = self.domain.element(vf)
```

Comment:
Here it's especially useful since you can give the vector field directly as a Numpy array of shape `(d, n1, ..., nd)` with `d = ndim`. I agree that this would be good to do in general.